cmake/scripts/gen_mingw_def.cmake

```diff
@@ -0,0 +1,114 @@
+set(tbb_32_FROM
```

Comment:
Such hardcoded symbols are fine to start with, but such approach will require manual adding new symbols here if they require similar replacements. I wonder if some universal rules can be applied, like `Kc` -> `Kw`, `Em` -> `Ey`, but seems they can't be reliable enough to work in 100% cases and do not touch unrelated parts.

Considering hardcoded symbols: I think there is not much value in comparison with just separate def files for MinGW, such automatic conversion may work worse in future development: developer has a set of def files in one place where he should add a new symbol, so he may hardly remember to add the symbol to the script in some other location.

Reply:
Like I said in my issue, the proper way to do this is to add the proper export attributes (`__declspec(dllexport)`, `__attribute__((visibility("default")))`) to the symbols in the source files itself. However, doing this can be time-consuming for such a large codebase.
Also, is there a specific need for the def files that the export attributes can't do?

Reply:
> the proper way to do this is to add export attributes (__declspec(dllexport), __attribute__((visibility("default")))) to the symbols in the source files itself

The Microsoft [documentation](https://docs.microsoft.com/en-us/cpp/build/determining-which-exporting-method-to-use?view=msvc-160) recommends using .def files because "_This is very convenient if you are designing a DLL for use by many applications because you can add new functionality and also ensure that it continues to work correctly with the applications that already rely on it_".

In addition, it warns about `__declspec(dllexport)`: "_If you rebuild the DLL with new exports, you also have to rebuild the applications because the decorated names for exported C++ functions might change if you use a different version of the compiler to rebuild it_".

Reply:
> they can't be reliable enough to work in 100% cases and do not touch unrelated parts.

I agree. We need to change `tchar`, `std::size_t`, `std::(u)intptr_t`, `std::ptrdiff_t` and somehow preserve actual `(unsigned) long`. With only the def files, this is not possible. Also there are many different combinations of parameters, and hardcoding all of them would be a pain.

Reply:
> In addition, it warns about `__declspec(dllexport)`: "_If you rebuild the DLL with new exports, you also have to rebuild the applications because the decorated names for exported C++ functions might change if you use a different version of the compiler to rebuild it_".

But the same situation would also happen with def files. If a new compiler version changes the name decorating rules, then the def files won't be correct anymore, and we'll have to maintain 2 files for the old and the new version, which in my opinion is worse than using `__declspec(dllexport)`. Thankfully, both the Itanium and Microsoft C++ ABI are now fairly stable, so we don't have to worry too much about symbol names over e.g. CRT mismatch.

> "This is very convenient if you are designing a DLL for use by many applications because you can add new functionality and also ensure that it continues to work correctly with the applications that already rely on it"

This is really only useful when we're exporting using ordinals, in which case it might be helpful to maintain compatibility with old DLLs. But considering that the library would be tied to the finished product, and also the fact that we don't provide prebuilt binaries, the chance of a DLL swap in a finished product is slim. Besides, we're exporting by names anyway, so it doesn't really matter. I haven't seen a case when `__declspec(dllexport)` actually fails.